### PR TITLE
compose: enable infra generation

### PR DESCRIPTION
**Pull request by weikanglim**
Enables infra generation, i.e. temporary infrastructure and synthesized infrastructure from resources.

Contributes to #4397,  https://github.com/Azure/azure-dev-pr/issues/1682

### CODE REVIEW

File: cli/azd/internal/scaffold/scaffold.go [hunk 1]
```diff
@@ -115,35 +116,94 @@ func ExecInfra(
...
+// are written to the target directory.
+func ExecInfraFs(
+	t *template.Template,
+	spec InfraSpec) (*memfs.FS, error) {
+	fs := memfs.New()
+
+	// Pre-execution expansion. Additional parameters are added, derived from the initial spec.
+	preExecExpand(&spec)
+
+	err := copyFsToMemFs(resources.ScaffoldBase, fs, baseRoot, ".")
```
Comment (vhvb1989) [inline cli/azd/internal/scaffold/scaffold.go:147]: image

ScaffoldBase is already a memory file system. Can you help me understand why do we need to copy it to a new memfs?

File: cli/azd/pkg/project/importer.go [hunk 2]
```diff
@@ -191,6 +198,10 @@ func (im *ImportManager) SynthAllInfrastructure(ctx context.Context, projectConf
 		}
 	}
 
+	if im.dotNetImporter.alphaFeatureManager.IsEnabled(alpha.FeatureId(alpha.Compose)) && len(projectConfig.Resources) > 0 {
+		return infraFsForProject(ctx, projectConfig)
+	}
+
 	return nil, fmt.Errorf("this project does not contain any infrastructure to synthesize")
```
Comment (vhvb1989) [inline cli/azd/pkg/project/importer.go:205]: If the alpha-feature is not enabled, are we printing this error message about there's not infra to synth?

File: cli/azd/pkg/project/resources.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	case ResourceTypeDbMongo:
+		return "MongoDB"
+	case ResourceTypeHostContainerApp:
+		return "Container App"
+	case ResourceTypeOpenAiModel:
+		return "Open AI Model"
+	}
+
+	return ""
+}
```
Comment (vhvb1989) [inline cli/azd/pkg/project/resources.go:37]: I'd use a `Resource Interface` here and create one file for each resource type.  Then make each type to implement `String() string`.
Then I'd create a Factory to define what Resource to create based on the name coming from azure.yaml.

I see a design like that easier to grow/evolve (even driven directly by the community) versus the current model where the more resources, the longer the `switch/case`, the longer this file.

Also for expanding what a Resource can do, It's easier to ask the editor to implement a new method in the interface in for all the implementations than asking to produce a new switch case for all cases in a new Method for `Resource`

**Review by vhvb1989 — APPROVED** [review]
Added a few comments. Nothing really blocking

File: cli/azd/pkg/project/resources.go (hunk 3 above)
Comment (weikanglim) [reply cli/azd/pkg/project/resources.go:37]: I don't disagree with the observation here that an interface is more loosely coupled than a strong union-type discriminator. It offers the flexibility at runtime that can be very useful for scenarios.

The trade-off we lose is compile-time safety. I'd like us to deeply think about the design here before committing to it.

The idea that "let's just use interfaces here" and "let's start with `Resource` interface", is indeed an artifact of how OOP philosophy is deeply ingrained in our current programming paradigm.

Adopting a go philosophy, I would start with thinking about the smallest interface we should expose.

Here's how K8s separates these ideas very cleanly for inspiration:

```golang
// Core interface that all Kubernetes runtime objects must implement
type Object interface {
    GetObjectKind() schema.ObjectKind
    DeepCopyObject() Object
}

// Interface for objects that can be stored in etcd
type ObjectMeta interface {
    GetName() string
    SetName(string)
    GetNamespace() string
    SetNamespace(string)
    GetLabels() map[string]string
    SetLabels(map[string]string)
    GetAnnotations() map[string]string
    SetAnnotations(map[string]string)
    GetResourceVersion() string
    SetResourceVersion(string)
}

// Interface for type information
type ObjectKind interface {
    SetGroupVersionKind(kind schema.GroupVersionKind)
    GroupVersionKind() schema.GroupVersionKind
}

// Interface for list types
type ListInterface interface {
    GetListMeta() ListMeta
    SetListMeta(ListMeta)
}

// Example of how a concrete type implements these interfaces
type Pod struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    Spec   PodSpec    `json:"spec,omitempty"`
    Status PodStatus  `json:"status,omitempty"`
}

// Implementation of Object interface
func (in *Pod) DeepCopyObject() runtime.Object {
    return in.DeepCopy()
}

// Client interface for pod operations
type PodInterface interface {
    Create(ctx context.Context, pod *v1.Pod, opts metav1.CreateOptions) (*v1.Pod, error)
    Update(ctx context.Context, pod *v1.Pod, opts metav1.UpdateOptions) (*v1.Pod, error)
    Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
    Get(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Pod, error)
    List(ctx context.Context, opts metav1.ListOptions) (*v1.PodList, error)
    Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error)
}
```

Over time, we can evolve it in a way that is more elegant (in my mind) than `Resource interface`

File: cli/azd/internal/scaffold/scaffold.go (hunk 1 above)
Comment (weikanglim) [reply cli/azd/internal/scaffold/scaffold.go:147]: `ScaffoldBase` is a filesystem containing assets to generate from.

Here, we are generating an in-memory filesystem contains the actual `infra` Bicep files that would be used for `azd provision`. This is how Aspire works today as well.

File: cli/azd/pkg/project/importer.go (hunk 2 above)
Comment (weikanglim) [reply cli/azd/pkg/project/importer.go:205]: We can definitely consider a better error message for enablement.

File: cli/azd/internal/scaffold/scaffold.go (hunk 1 above)
Comment (vhvb1989) [reply cli/azd/internal/scaffold/scaffold.go:147]: I'm confused about the copy from fs to a new fs.
We create the template directly from the fs with `ParseFS()` https://github.com/Azure/azure-dev/blob/main/cli/azd/pkg/apphost/generate.go#L99

Then executeToFS() uses an existing fs.   Is this `existing fs` the equivalent to the new fs you are creating?

That's why I wonder if you can just re-use the same fs, maybe just to use a root folder like `out` ?

File: cli/azd/pkg/project/scaffold_gen.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+		if isSecret {
+			// TODO: handle secrets
+			continue
+		}
+
+		svcSpec.Env[envVar.Name] = value
+	}
+
+	port := props.Port
+	if port < 1 || port > 65535 {
```
Comment (wbreza) [inline cli/azd/pkg/project/scaffold_gen.go:203]: nit: consider moving 65535 out to a constant.

**Review by wbreza — APPROVED** [review]

File: cli/azd/pkg/project/scaffold_gen.go (hunk 4 above)
Comment (weikanglim) [reply cli/azd/pkg/project/scaffold_gen.go:203]: I will leave it as-is for now -- we could think about creating some sort of a networking package that validates port numbers amongst other things. But creating a constant right now would end up adding to global variables in `project` that wouldn't quite be consumable either, so I'm pretty happy with how it looks right now.

**Pull request merged**